File: Awful.apk/src/main/java/com/ferg/awfulapp/task/BookmarkRequest.java
package com.ferg.awfulapp.task;

import android.content.ContentValues;
import android.content.Context;
import android.net.Uri;
import com.ferg.awfulapp.constants.Constants;
import com.ferg.awfulapp.provider.AwfulProvider;
import com.ferg.awfulapp.thread.AwfulThread;
import com.ferg.awfulapp.util.AwfulError;
import org.jsoup.nodes.Document;

/**
 * Created by matt on 8/8/13.
 */
public class BookmarkRequest extends AwfulRequest<Void> {
    private int threadId;
    private boolean add;
    public BookmarkRequest(Context context, int threadId, boolean add) {
        super(context, null);
        this.add = add;
        this.threadId = threadId;
    }

    @Override
    protected String generateUrl(Uri.Builder urlBuilder) {
        addPostParam(Constants.PARAM_THREAD_ID, Integer.toString(threadId));
        if(add){
            addPostParam(Constants.PARAM_ACTION, "add");
        }else{
            addPostParam(Constants.PARAM_ACTION, "remove");
        }
        return Constants.FUNCTION_BOOKMARK;
    }

    @Override
    protected Void handleResponse(Document doc) throws AwfulError {
        ContentValues cv = new ContentValues();
        cv.put(AwfulThread.BOOKMARKED, add?1:0);
        getContentResolver().update(AwfulThread.CONTENT_URI, cv, AwfulThread.ID+"=?", AwfulProvider.int2StrArray(threadId));
        if(!add){
            getContentResolver().delete(AwfulThread.CONTENT_URI_UCP, AwfulThread.ID+"=?", AwfulProvider.int2StrArray(threadId));
        }
        return null;
    }

    @Override
    protected boolean handleError(AwfulError error, Document doc) {
        return error.isCritical();
    }
}

File: Awful.apk/src/main/java/com/ferg/awfulapp/task/BookmarkRequest.kt
package com.ferg.awfulapp.task

import android.content.ContentValues
import android.content.Context
import com.ferg.awfulapp.constants.Constants.*
import com.ferg.awfulapp.thread.AwfulThread
import com.ferg.awfulapp.util.AwfulError
import com.ferg.awfulapp.util.toSqlBoolean
import org.jsoup.nodes.Document

/**
 * Add or remove a bookmark on the site for the given [threadId], updating the local database.
 */
class BookmarkRequest(context: Context, private val threadId: Int, private val add: Boolean)
    : AwfulRequest<Void?>(context, FUNCTION_BOOKMARK, isPostRequest = true) {

    init {
        with(parameters) {
            add(PARAM_THREAD_ID, threadId.toString())
            add(PARAM_ACTION, if (add) "add" else "remove")
        }
    }


    @Throws(AwfulError::class)
    override fun handleResponse(doc: Document): Void? {
        val cv = ContentValues()
        cv.put(AwfulThread.BOOKMARKED, add.toSqlBoolean)
        val id = arrayOf(threadId.toString())

        with(contentResolver) {
            update(AwfulThread.CONTENT_URI, cv, "${AwfulThread.ID}=?", id)
            if (!add) delete(AwfulThread.CONTENT_URI_UCP, "${AwfulThread.ID}=?", id)
        }
        return null
    }

}
